kernel/rtlil.h

```diff
@@ -657,38 +665,205 @@ namespace RTLIL
 
 	std::string decode_string() const;
+	inline int size() const {
+		if (is_str())
+			return 8 * str_.size();
+		else {
+			log_assert(is_bits() && "malformed Const union");
+			return bits_.size();
+		}
+	}
+
+	inline bool empty() const {
+		if (is_str())
+			return str_.empty();
+		else {
+			log_assert(is_bits() && "malformed Const union");
+			return bits_.empty();
+		}
+	}
+
+	void bitvectorize() const {
+		if (tag == backing_tag::bits)
+			return;
+
+		log_assert(is_str() && "malformed Const union");
+
+		bitvectype new_bits;
+
+		new_bits.reserve(str_.size() * 8);
+		for (int i = str_.size() - 1; i >= 0; i--) {
+			unsigned char ch = str_[i];
+			for (int j = 0; j < 8; j++) {
+				new_bits.push_back((ch & 1) != 0 ? State::S1 : State::S0);
+				ch = ch >> 1;
+			}
+		}
+
+		{
+			// sketchy zone
+			str_.~string();
+			(void)new ((void*)&bits_) bitvectype(std::move(new_bits));
+			tag = backing_tag::bits;
+		}
+	}
+
+	inline RTLIL::State &operator[](int index) {
+		return bits().at(index);
+	}
+	inline RTLIL::State operator[](int index) const {
+		return bits().at(index);
+	}
+	inline bitvectype::iterator begin() {
+		return bits().begin();
+	}
+	inline bitvectype::iterator end() {
+		return bits().end();
+	}
+
+	class const_bit_view {
+	private:
+		const Const& parent;
+	public:
+		const_bit_view(const Const& c) : parent(c) {}
+
+		class const_iterator {
+		private:
+			const Const& parent;
+			size_t idx;
 
-	inline int size() const { return bits.size(); }
-	inline bool empty() const { return bits.empty(); }
-	inline RTLIL::State &operator[](int index) { return bits.at(index); }
-	inline const RTLIL::State &operator[](int index) const { return bits.at(index); }
-	inline decltype(bits)::iterator begin() { return bits.begin(); }
-	inline decltype(bits)::iterator end() { return bits.end(); }
+		public:
+			using iterator_category = std::input_iterator_tag;
+			using value_type = State;
+			using difference_type = std::ptrdiff_t;
+			using pointer = const State*;
+			using reference = State;
+
+			const_iterator(const Const& c, size_t i) : parent(c), idx(i) {}
+
+			State operator*() const {
+				if (auto bv = parent.get_if_bits())
+					return (*bv)[idx];
+
+				bool bit = (parent.get_str()[idx / 8] & (1 << (7 - (idx % 8))));
+				return bit ? State::S1 : State::S0;
+			}
+			const_iterator& operator++() {
+				++idx;
+				return *this;
+			}
+			const_iterator& operator--() {
+				--idx;
+				return *this;
+			}
+			const_iterator& operator++(int) {
+				++idx;
+				return *this;
+			}
+			const_iterator& operator--(int) {
+				--idx;
+				return *this;
+			}
+			const_iterator& operator+=(int i) {
+				idx += i;
+				return *this;
+			}
+
+			const_iterator operator+(int add) {
+				return const_iterator(parent, idx + add);
+			}
+			int operator-(const const_iterator& other) {
+				return idx - other.idx;
+			}
+
+			bool operator==(const const_iterator& other) const {
+				return idx == other.idx;
```

Comment:
`&& &parent == &other.parent`, or are comparisons on iterators from different parents undefined? Not sure here what is mandated